src/examples/4626-router/Generic4626Router.sol

```diff
@@ -0,0 +1,229 @@
+        bytes calldata
+    )
+        internal
+        override
+        returns (bytes4 selector, BeforeSwapDelta swapDelta, uint24 lpFeeOverride)
+    {
+        PoolId poolId = poolKey.toId();
+        PoolDetails memory details = poolDetails[poolId];
+
+        bool isExactInput = params.amountSpecified < 0;
+
+        // This following section is hard to follow, I initially wrote this with currency0 and currency1
+        // but it was truly impossible to follow. Ended up storing the vault and underlying.
+        // Although, I think I got a hang of it so might go back
+        if (params.zeroForOne == details.wrapDirection) {
+            uint256 inputAmount = isExactInput
+                ? uint256(-params.amountSpecified)
+                : _getUnderlyingForShares(details.vault, uint256(params.amountSpecified));
+
+            CurrencySettler.take(
+                Currency.wrap(address(details.underlying)),
+                poolManager,
+                address(this),
+                inputAmount,
+                false
+            );
+
+            uint256 shares = _deposit(details.underlying, details.vault, inputAmount);
+
+            CurrencySettler.settle(
+                Currency.wrap(address(details.vault)), poolManager, address(this), shares, false
+            );
+
+            int128 amountUnspecified =
+                isExactInput ? -shares.toInt256().toInt128() : inputAmount.toInt256().toInt128();
+
+            swapDelta = toBeforeSwapDelta(-params.amountSpecified.toInt128(), amountUnspecified);
+        } else {
+            uint256 inputAmount = isExactInput
+                ? uint256(-params.amountSpecified)
+                : _getSharesForUnderlying(details.vault, uint256(params.amountSpecified));
+
+            CurrencySettler.take(
+                Currency.wrap(address(details.vault)),
+                poolManager,
+                address(this),
+                inputAmount,
+                false
+            );
+
+            uint256 underlyingAmount = _withdraw(details.vault, inputAmount);
+
+            CurrencySettler.settle(
+                Currency.wrap(address(details.underlying)),
+                poolManager,
+                address(this),
+                underlyingAmount,
+                false
+            );
+
+            int128 amountUnspecified = isExactInput
+                ? -underlyingAmount.toInt256().toInt128()
+                : inputAmount.toInt256().toInt128();
+
+            swapDelta = toBeforeSwapDelta(-params.amountSpecified.toInt128(), amountUnspecified);
+        }
+
+        return (IHooks.beforeSwap.selector, swapDelta, 0);
+    }
+
+    function _deposit(
+        ERC20 underlying,
+        ERC4626 vault,
+        uint256 underlyingAmount
+    ) internal returns (uint256 shares) {
+        // TODO: Is it worth protecting from lingering approvals, or just approve max in initializer?
+        underlying.approve(address(vault), underlyingAmount);
```

Comment:
max approve in initializer is kind of safe. its safe strictly because the hook never holds the ERC20 balance - we should document it though. if someone were to fork this code and hold an ERC20 balance, malicious vaults might be able to transferFrom hook to attacker

also some tokens will decrement the approvals so the `_deposit` function should check if the allowance is insufficient and max approve again